client/beefy/src/worker.rs

```diff
@@ -430,6 +444,19 @@ where
 	}
 }
 
+/// Returns a future that waits until major sync is done
+/// before completing
+fn wait_for_major_syncing<SO: SyncOracle + Send + Sync + Clone + 'static>(
+	mut sync_oracle: SO,
+) -> impl future::Future<Output = ()> {
+	return future::poll_fn(move |_cx| {
+		if sync_oracle.is_major_syncing() {
+			Poll::Pending
+		} else {
+			Poll::Ready(())
+		}
+	})
```

Comment:
this function has no way of knowing when the sync is over, i suggest adding it to the  sync_oracle itself

Reply:
What if it's rescheduled for polling with `cx.waker().wake_by_ref()` before returning pending, Would that be fine?